### ISSUE TITLE
X-Request-Id Middleware

### ISSUE THREAD

**Issue by Slakah** [http4s/http4s#3485]
We've got an internally used middleware, and wondering whether this would be useful to contribute back into http4s. The basic concept is, propagate the header from request to response, generating a random UUID `X-Request-Id` if it wasn't set.

Gist of the code (older version of http4s as can be seen by presence of `.ci`): https://gist.github.com/Slakah/f5eb4fbe3c6cc1ed28f8bd66482cd566

Docs:
* [Heroku docs on X-Request-Id](https://devcenter.heroku.com/articles/http-request-id)
* [Stackoverflow](https://stackoverflow.com/questions/25433258/what-is-the-x-request-id-http-header)

Does it also make sense to add official support for the header as well in the [`headers` package](https://github.com/http4s/http4s/tree/master/core/src/main/scala/org/http4s/headers), i.e.

```
package org.http4s
package headers

object `X-Requested-With` extends HeaderKey.Default
```

Let me know if this is something you'll be interested in, and I can draft a PR.

**Comment by hamnis** [comment]
Option 2 is reasonable.

**Comment by Slakah** [comment]
Implemented by #3487
